File: go/sort-in-one-loop.go
package main

import (
	"fmt"
	"math/rand"
)

// Simple trick to do array sort in just one loop
func sort_in_one_loop() {
	fmt.Println("sort in one loop")

	var array_number [50]int
	array_length := len(array_number)

	for i := 0; i < array_length; i++ {
		array_number[i] = rand.Intn(100)
	}

	fmt.Println("random number: ", array_number)

	i := 0
	for i < array_length {

		if i != array_length-1 &&
			i >= 0 &&
			array_number[i] > array_number[i+1] {
			x := array_number[i]
			array_number[i] = array_number[i+1]
			array_number[i+1] = x
			i -= 2
		}

		i++
	}

	fmt.Println("sorted number: ", array_number)
}

File: c/sort/sort-in-one-loop.c
#include <stdio.h>
#include <stdlib.h>

void sort_in_one_loop()
{
  printf("sort in one loop\n");

  int array_length = 50;
  int array_number[array_length];
  int i;
  int x;

  printf("random number:");
  for (i = 0; i < array_length; i++)
  {
    array_number[i] = rand() % 10;
    printf(" %d", array_number[i]);
  }
  printf("\n");

  i = 0;

  while (i < array_length)
  {
    if (i != (array_length - 1) && i >= 0 && array_number[i] > array_number[i + 1])
    {
      x = array_number[i];
      array_number[i] = array_number[i + 1];
      array_number[i + 1] = x;
      i -= 2;
    }
    i++;
  }

  printf("sorted number: ");
  for (i = 0; i < array_length; i++)
  {
    printf(" %d", array_number[i]);
  }
  printf("\n");
}
